inc/Engine/CriticalPath/RESTWP.php

```diff
@@ -0,0 +1,480 @@
+<?php
+
+namespace WP_Rocket\Engine\CriticalPath;
+
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_Error;
+
+/**
+ * Class RESTWP
+ *
+ * @package WP_Rocket\Engine\CriticalPath
+ */
+abstract class RESTWP {
+
+	/**
+	 * Namespace for REST Route.
+	 */
+	const ROUTE_NAMESPACE = 'wp-rocket/v1';
+
+	/**
+	 * Part of route namespace for this inherited class item type.
+	 *
+	 * @var string $route_namespace to be set with like post, term.
+	 */
+	protected $route_namespace;
+
+	/**
+	 * Responsible for dealing with data/database.
+	 *
+	 * @var DataManager datamanager instance.
+	 */
+	private $data_manager;
+
+	/**
+	 * Responsible for dealing with CPCSS APIs.
+	 *
+	 * @var APIClient api_client instance.
+	 */
+	private $api_client;
+
+	/**
+	 * RESTWP constructor.
+	 *
+	 * @since 3.6
+	 *
+	 * @param DataManager $data_manager Data manager instance, responsible for dealing with data/database.
+	 * @param APIClient   $api_client API Client instance to deal with CPCSS APIs.
+	 */
+	public function __construct( DataManager $data_manager, APIClient $api_client ) {
+		$this->data_manager = $data_manager;
+		$this->api_client   = $api_client;
+	}
+
+	/**
+	 * Registers the generate route in the WP REST API
+	 *
+	 * @since 3.6
+	 *
+	 * @return void
+	 */
+	public function register_generate_route() {
+		register_rest_route(
+			self::ROUTE_NAMESPACE,
+			'cpcss/' . $this->route_namespace . '/(?P<id>[\d]+)',
+			[
+				'methods'             => 'POST',
+				'callback'            => [ $this, 'generate' ],
+				'permission_callback' => [ $this, 'check_permissions' ],
+			]
+		);
+	}
+
+	/**
+	 * Register Delete CPCSS route in the WP REST API.
+	 *
+	 * @since  3.6
+	 */
+	public function register_delete_route() {
+		register_rest_route(
+			self::ROUTE_NAMESPACE,
+			'cpcss/' . $this->route_namespace . '/(?P<id>[\d]+)',
+			[
+				'methods'             => 'DELETE',
+				'callback'            => [ $this, 'delete' ],
+				'permission_callback' => [ $this, 'check_permissions' ],
+			]
+		);
+	}
```

Comment:
This should be part of `RESTWPPost` instead, it's specific to the posts generation. Let's keep `RESTWP` as generic as possible